### PR TITLE
Refactor Product service API

### PR DESCRIPTION
**Pull request by sav007**
Change collectionId parameter for getProducts method mandatory
Add new api call to get products filtered by tags only

@bgulanowski @krisorr @davidmuzi

### CODE REVIEW

File: MobileBuy/buy/src/androidTest/java/com/shopify/buy/service/StorefrontTest.java [hunk 1]
```diff
@@ -210,7 +211,16 @@ public void testGetCollectionPage() throws InterruptedException {
             public void success(List<Collection> collections) {
                 assertNotNull(collections);
                 assertFalse(collections.isEmpty());
-                assertEquals(collections.get(0).getHandle(), "frontpage");
+
+                boolean found = false;
+                for (Collection collection : collections) {
+                    if ("frontpage".equals(collection.getHandle())) {
```
Comment (krisorr) [inline MobileBuy/buy/src/androidTest/java/com/shopify/buy/service/StorefrontTest.java]: should be a define in the test data.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java [hunk 2]
```diff
@@ -82,7 +82,7 @@ public CancellableTask getProducts(final int page, final Callback<List<Product>>
 
     @Override
     public Observable<List<Product>> getProducts(final int page) {
-        return getProducts(page, null, null, null);
+        return getProducts(page, (Set<String>) null);
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: Why not null here?  seems redundant to make an empty set.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java [hunk 3]
```diff
@@ -229,6 +229,27 @@ public CancellableTask getProductTags(int page, Callback<List<String>> callback)
...
+        return new CancellableTaskSubscriptionWrapper(getProducts(page, tags).subscribe(new InternalCallbackSubscriber<>(callback)));
+    }
+
+    @Override
+    public Observable<List<Product>> getProducts(final int page, final Set<String> tags) {
+        if (page < 1) {
+            throw new IllegalArgumentException("page is a 1-based index, value cannot be less than 1");
+        }
+
+        final String tagsQueryStr = tags != null && !tags.isEmpty() ? TextUtils.join(",", tags.toArray()) : null;
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: I would prefer to see this logic broken up into a couple of lines, it is a bit hard to parse.
As well, then you could avoid the need to create the empty set to pass in for the other getProducts(page) call.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java [hunk 4]
```diff
@@ -239,13 +260,12 @@ public CancellableTask getProducts(final int page, final Long collectionId, Set<
...
         }
 
-        String sortOrderStr = sortOrder != null ? sortOrder.toString() : Collection.SortOrder.COLLECTION_DEFAULT.toString();
         if (collectionId == null) {
-            sortOrderStr = null;
+            throw new NullPointerException("collectionId cannot be null");
         }
 
+        final String sortOrderStr = sortOrder != null ? sortOrder.toString() : Collection.SortOrder.COLLECTION_DEFAULT.toString();
         final String tagsQueryStr = tags != null && !tags.isEmpty() ? TextUtils.join(",", tags.toArray()) : null;
```
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: We do the same ternary operator above - how about breaking this out into a seperate function.

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java (hunk 2 above)
Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: ambiguous call error, we have another method with 2 params

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java (hunk 3 above)
Comment (sav007) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: Sorry confused: "As well, then you could avoid the need to create the empty set to pass in for the other getProducts(page) call." where do you see the call with empty set?

File: MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java (hunk 2 above)
Comment (krisorr) [inline MobileBuy/buy/src/main/java/com/shopify/buy/dataprovider/ProductServiceDefault.java]: okay, fair enough

**Comment by krisorr** [comment]
LGTM
